packages/core/src/errors.ts

```diff
@@ -1,58 +1,176 @@
+
 /**
- * All exceptions intenionally thrown with Ignition-core
- * extend this class.
+ * Base error class extended by all custom errors.
+ * Placeholder to allow us to customize error output formatting in the future.
  *
  * @alpha
  */
-export class IgnitionError extends Error {
+export class CustomError extends Error {
   constructor(message: string) {
     super(message);
+    this.name = this.constructor.name;
+  }
+}
+
+/**
+ * All exceptions intentionally thrown with Ignition-core
+ * extend this class.
+ *
+ * @alpha
+ */
+export class IgnitionError extends CustomError {
+  constructor(
+    errorDescriptor: ErrorDescriptor,
+    messageArguments: Record<string, string | number> = {}
+  ) {
+    const prefix = `${getErrorCode(errorDescriptor)}: `;
+    const formattedMessage = applyErrorMessageTemplate(
+      errorDescriptor.message,
+      messageArguments
+    );
+
+    super(prefix + formattedMessage);
 
     this.name = this.constructor.name;
   }
 }
 
 /**
- * This error class represents issue detected by Ignition-cores
- * validation phase on the user inputed module. Validation errors
- * capture the stack to the action within the offending module,
- * to enhance the locality of the validation error message.
+ * This class is used to throw errors from Ignition plugins made by third parties.
  *
- * * @alpha
+ * @alpha
  */
-export class IgnitionValidationError extends IgnitionError {
-  constructor(message: string) {
+export class IgnitionPluginError extends CustomError {
+  public static isIgnitionPluginError(
+    error: any
+  ): error is IgnitionPluginError {
+    return (
+      typeof error === "object" &&
+      error !== null &&
+      (error as IgnitionPluginError)._isIgnitionPluginError === true
+    );
+  }
+
+  private readonly _isIgnitionPluginError = true;
+
+  public readonly pluginName: string;
+
+  constructor(pluginName: string, message: string) {
     super(message);
+    this.pluginName = pluginName;
+    Object.setPrototypeOf(this, IgnitionPluginError.prototype);
+  }
+}
 
-    // This is required to allow calls to `resetStackFrom`,
-    // otherwise the function is not available on the
-    // error instance
-    Object.setPrototypeOf(this, new.target.prototype);
-  }
-
-  /**
-   * Reset the stack hiding parts that are bellow the given function.
-   * The intention is the function should be part of the user callable
-   * api, so that the stack leads directly to the line in the module
-   * the user called (i.e. `m.contract(...)`)
-   *
-   * @param f - the function to hide all of the stacktrace above
-   *
-   * @internal
-   */
-  public resetStackFrom(f: () => any) {
-    Error.captureStackTrace(this, f);
+/**
+ * This class is used to throw errors from *core* Ignition plugins.
+ * If you are developing a third-party plugin, use IgnitionPluginError instead.
+ *
+ * @alpha
+ */
+export class NomicIgnitionPluginError extends IgnitionPluginError {
```

Comment:
I added this and the above `IgnitionPluginError` again mimicking the similar classes from Hardhat.